invenio_rdm_migrator/state.py

```diff
@@ -316,22 +326,39 @@ def search(self, column, value):
 
     def all(self):
         """Get all rows."""
-        for row in self.state.all(self.table_name):
-            yield self._row_as_dict(row)
+        if self._cache is not None:
+            for v in self._cache.values():
+                yield v
+        else:
+            for row in self.state.all(self.table_name):
+                yield self._row_as_dict(row)
 
     def add(self, key, data):
         """Add data row."""
-        self.state.add(self.table_name, {self.pk_attr: key, **data})
+        self.state.validate(self.table_name, data)
+        if self._cache is not None:
+            if key in self._cache:
+                raise ValueError("Key {key} already in state.")
+            self._cache[key] = data
```

Comment:
> * Kafka/Incremental why is okay

For Kafka/incremental we need persistence and actually maybe even use disk persistence directly, since there's no clear point while processing the stream to decide on saving to disk... Possibly we can keep the memory cache as well as an L1 cache.

> * We would need to somehow go back to how the DB was after the snapshot migration. Do we have a plan for that?

Yeah, I would like to dump the in-memory cache actually at the end of each stream (like we do now). Currently though, the `self._cache` does not get dumped back to the main `state_db`

Reply:
> For Kafka/incremental we need persistence and actually maybe even use disk persistence directly, since there's no clear point while processing the stream to decide on saving to disk... Possibly we can keep the memory cache as well as an L1 cache.

In some placed, I updated the cache (DB). So there would be no need to dump it per-se. It would be using the latest. It's true this could end up in a corrupeted state.

Reply:
I was thinking that "state" should be committed inside a transaction, probably...